File: Python/7py 2/halving_sum.py
"""
|--------------------------------------------------------------------------
| Halving Sum
|--------------------------------------------------------------------------
|
"""

from math import floor
def halving_sum(n):
   return 1 if n == 1 else n + halving_sum(floor(n/2))


def halving_sum(n): 
   return n and n + halving_sum(n>>1)


def halving_sum(n):
   result = n
   while n >= 1:
      n //= 2
      result += n
   return result


def halving_sum(n):
   return n + halving_sum(n // 2) if n else 0


print(halving_sum(25),47)
print(halving_sum(127),247)

File: JavaScript/7js 1/halvingSum.js
/*
|--------------------------------------------------------------------------
| Halving Sum
|--------------------------------------------------------------------------
|
*/

function halvingSum(n) {
   let res = n
   while (Math.floor(n / 2) > 0) {
      res += Math.floor(n / 2);
      n = Math.floor(n / 2);
   }
   return res;
}


function halvingSum(n) {
   return n === 1 ? 1 : n + halvingSum(Math.floor(n / 2))
}


const halvingSum = n => n + (n > 1 && halvingSum(n >> 1));


function halvingSum(n) {
   let sum = 0;
   while (n > 0) {
      sum += n;
      n = Math.floor(n / 2);
   }
   return sum;
}


function halvingSum(n) {
   return n * 2 - eval([...n.toString(2)].join `+`)
}


console.log(halvingSum(25), 47);
console.log(halvingSum(127), 247);